File: codility/go/7-stacks_queues.go
package golang

func brackets(S string) int {
	if len(S) == 0 {
		return 1
	}

	brackets := make([]string, 0, len(S))
	for _, char := range S {
		b := string(char)
		if b == "(" || b == "[" || b == "{" {
			// we can always push opening brackets
			brackets = append(brackets, b)
			continue
		}

		// it's a closing bracket. check if last opening bracket is the same type
		if len(brackets) == 0 {
			return 0
		}

		last := brackets[len(brackets)-1]
		if (b == ")" && last == "(") || (b == "]" && last == "[") || (b == "}" && last == "{") {
			brackets = brackets[:len(brackets)-1]
			continue
		}

		// closing bracket doesn't match last opening bracket
		return 0
	}

	if len(brackets) > 0 {
		return 0
	}

	return 1
}

func fish(A []int, B []int) int {
	// stack the fish going downstream and have them eat smaller fish going upstream as they pass,
	// or get eaten by bigger fish going downstream
	// count the fish that survive
	var downstream []int
	var alive int
	for i := 0; i < len(A); i++ {
		// if going downstream, just add to the stack
		if B[i] == 1 {
			downstream = append(downstream, A[i])
			continue
		}

		// going upstream. eat all smaller downstream fish
		for len(downstream) > 0 {
			if A[i] > downstream[len(downstream)-1] {
				downstream = downstream[:len(downstream)-1]
				continue
			}

			// last downstream fish is bigger
			break
		}

		// if there are no downstream fish left, the upstream fish survives
		if len(downstream) == 0 {
			alive++
		}
	}

	return alive + len(downstream)
}

func nesting(S string) int {
	open := make([]string, 0, len(S)/2+1)
	for _, char := range S {
		s := string(char)
		if s == "(" {
			open = append(open, s)
			continue
		}
		// char is ")"
		if len(open) == 0 {
			return 0
		}
		if open[len(open)-1] != "(" {
			return 0
		}
		open = open[:len(open)-1]
	}

	if len(open) == 0 {
		return 1
	}

	return 0
}

func stoneWall(H []int) int {
	var wall []int
	var count int
	for _, h := range H {
		// remove all higher previous stones, going backwards, then count next stone if wall is empty or last stone is lower
		for ii := len(wall) - 1; ii >= 0; ii-- {
			if wall[ii] < h {
				break
			}
			if wall[ii] > h {
				wall = wall[:ii]
			}
		}

		if len(wall) == 0 || wall[len(wall)-1] < h {
			wall = append(wall, h)
			count++
		}
	}

	return count
}

File: codility/py/7-stacks_queues.py

def brackets(S):
    if not S:
        return 1

    if len(S) & 1:
        return 0

    stack = []
    for c in S:
        if c in '({[':
            stack.append(c)
        else:
            if not stack:
                return 0
            top = stack.pop()
            if (c == ')' and top != '(') or (c == '}' and top != '{') or (c == ']' and top != '['):
                return 0

    return 1 if not stack else 0

assert brackets("") == 1  # empty string
assert brackets("()") == 1
assert brackets("[]") == 1
assert brackets("{}") == 1
assert brackets("{[()()]}") == 1
assert brackets("(([]){})") == 1
assert brackets("({[]})") == 1
assert brackets("[{()}](){}") == 1
assert brackets("(") == 0
assert brackets("]") == 0
assert brackets("([)]") == 0
assert brackets("({[})]") == 0
assert brackets("(()") == 0
assert brackets("())") == 0
assert brackets("}") == 0       # closing bracket with no opener
assert brackets("{") == 0       # opening bracket only
assert brackets("}{") == 0      # reversed closing then opening
assert brackets("[({})](]") == 0  # almost valid but one mismatch at end
assert brackets("}") == 0       # closing bracket with no opener
assert brackets("{") == 0       # opening bracket only
assert brackets("}{") == 0      # reversed closing then opening
assert brackets("[({})](]") == 0  # almost valid but one mismatch at end

def fish(A,B):
    downstream,survived = [], 0
    for i in range(len(A)):
        size, dir = A[i],B[i]
        # Move downstream fish to their stack
        if dir == 1:
            downstream.append(size)
        else:
            # Eat all smaller downstream fish
            while downstream and downstream[-1] < size:
                downstream.pop()
            # If all eaten, upstream fish survives
            if not downstream:
                survived += 1

    # Total survivors include survived downstream fish amd and recorded count of upstream survived fish
    return len(downstream) + survived

def nesting(S):
    open_params = 0
    for c in S:
        if s == "(":
            open_params += 1
        elif open_params:
            open_params -= 1
        else:
            return 0
    # open_params needs to be empty
    return 1 if not open_params else 0
